engine/src/main/java/io/camunda/zeebe/engine/processing/message/MessageTimeToLiveChecker.java

```diff
@@ -7,36 +7,90 @@
+ * MessageObserver#MESSAGE_TIME_TO_LIVE_CHECK_INTERVAL interval}. If it reschedules itself
+ * immediately, then it will continue where it left off the last time. Otherwise, it starts with the
+ * first expired message deadline it can find.
+ */
 public final class MessageTimeToLiveChecker implements Task {
 
   private static final MessageRecord EMPTY_DELETE_MESSAGE_COMMAND =
       new MessageRecord().setName("").setCorrelationKey("").setTimeToLive(-1L);
 
+  private static final int EXPIRE_COMMANDS_LIMIT_DEFAULT = 10;
+
+  /** This determines the maximum number of EXPIRE commands it will attempt to fit in the result. */
+  private static final int LIMIT = EXPIRE_COMMANDS_LIMIT_DEFAULT;
+
+  private final ProcessingScheduleService scheduleService;
   private final MessageState messageState;
 
-  public MessageTimeToLiveChecker(final MessageState messageState) {
+  /** Keeps track of the timestamp to compare the message deadlines against. */
+  private long currentTimestamp = -1;
+  /** Keeps track of where to continue between iterations. */
+  private MessageState.Index lastIndex;
+
+  public MessageTimeToLiveChecker(
+      final ProcessingScheduleService scheduleService, final MessageState messageState) {
+    this.scheduleService = scheduleService;
     this.messageState = messageState;
+    lastIndex = null;
   }
 
   @Override
   public TaskResult execute(final TaskResultBuilder taskResultBuilder) {
-    messageState.visitMessagesWithDeadlineBefore(
-        ActorClock.currentTimeMillis(),
-        expiredMessageKey -> writeDeleteMessageCommand(expiredMessageKey, taskResultBuilder));
-    return taskResultBuilder.build();
-  }
+    if (currentTimestamp == -1) {
+      currentTimestamp = ActorClock.currentTimeMillis();
+    }
 
-  private boolean writeDeleteMessageCommand(
-      final long expiredMessageKey, final TaskResultBuilder taskResultBuilder) {
-    return taskResultBuilder.appendCommandRecord(
-        expiredMessageKey, MessageIntent.EXPIRE, EMPTY_DELETE_MESSAGE_COMMAND);
+    final var counter = new MutableInteger(0);
+    final boolean shouldContinueWhereLeftOff =
+        messageState.visitMessagesWithDeadlineBeforeTimestamp(
+            currentTimestamp,
+            lastIndex,
+            (deadline, expiredMessageKey) -> {
+              final var newIndex = new Index(expiredMessageKey, deadline);
+              final boolean wasIndexAlreadyVisitedLastTime = newIndex.equals(lastIndex);
+              lastIndex = newIndex;
+
+              if (wasIndexAlreadyVisitedLastTime) {
+                // skip this entry
+                return true;
+              }
+
+              final boolean stillFitsInResult =
+                  taskResultBuilder.appendCommandRecord(
+                      expiredMessageKey, MessageIntent.EXPIRE, EMPTY_DELETE_MESSAGE_COMMAND);
+              return stillFitsInResult && counter.incrementAndGet() <= LIMIT;
```

Comment:
:wrench: Sorry, I missed mentioning it in the first review round. With `<= LIMIT`, it will collect at max `11` (`=LIMIT + 1` ). So strictly speaking, it is not `10` (`=LIMIT`) as the default limit suggests. However, it is nothing blocking to merge this PR, just wanted to mention it for completeness.

```suggestion
              return stillFitsInResult && counter.incrementAndGet() < LIMIT;
```

Reply:
Good catch! Easy to fix!